### PR TITLE
fix: rename keep_bindings to keep_vars, and make it opt-in, to keep wrangler.toml compatible with being used for IaC

### LINKED ISSUE #1897: 🐛 BUG: wrangler keeps all vars given across dashboard and wrangler.toml, breaking IaC
### What version of `Wrangler` are you using?

2.1.5

### What operating system are you using?

Mac

### Describe the Bug

It should be possible to use wrangler.toml as the source of truth by default.

Due to the way we've implemented `keep_bindings`, this is no longer the case.

### PR DESCRIPTION
**Pull request by rozenmd**
Closes #1897

### CODE REVIEW

File: packages/wrangler/src/index.tsx [hunk 1]
```diff
@@ -527,6 +527,12 @@ function createCLIParser(argv: string[]) {
 						describe: "Don't actually publish",
 						type: "boolean",
 					})
+					.option("keep-vars", {
+						describe:
+							"Stop wrangler from deleting vars that are not present in the wrangler.toml",
```
Comment (petebacondarwin) [inline packages/wrangler/src/index.tsx:532]: I think this needs to be expanded a bit more.

**Review by JacobMGEvans — APPROVED** [review]
I think as long as the behavior is the config by default is the source of truth than this is a great addition, especially considering recent additions like `--from-dash`

File: packages/wrangler/src/config/config.ts [hunk 2]
```diff
@@ -168,6 +168,18 @@ export interface ConfigFields<Dev extends RawDevConfig> {
...
+	 * By default, wrangler.toml is the source of truth for your environment configuration, like a terraform file.
+	 *
+	 * If you change your vars in the dashboard, wrangler *will* override/delete them on its next publish.
+	 *
+	 * If you want to keep your dashboard vars when wrangler deploys, set this field to true.
+	 *
+	 * @default false
+	 * @nonInheritable
+	 */
+	keep_vars?: boolean;
```
Comment (petebacondarwin) [inline packages/wrangler/src/config/config.ts:182]: Adding this type here implies that it is possible to set this via wrangler.toml.
Is that what you intend? If so then we need to add code to `validation.ts` to check it and copy if over; and also tests in `configuration.test.ts`.

File: packages/wrangler/src/publish.ts [hunk 3]
```diff
@@ -502,7 +504,7 @@ See https://developers.cloudflare.com/workers/platform/compatibility-dates for m
 			compatibility_flags:
 				props.compatibilityFlags ?? config.compatibility_flags,
 			usage_model: config.usage_model,
-			keep_bindings: true,
+			keep_bindings: keepVars,
```
Comment (petebacondarwin) [inline packages/wrangler/src/publish.ts:507]: Given that this is a boolean then I think we should actually rename this property to `keepVars` too.
We should only switch to `keep_bindings` at the point when we convert from boolean to the array of bindings to keep.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
We need to create an issue to track documenting this.

File: packages/wrangler/src/publish.ts (hunk 3 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/publish.ts:507]: Ok! I was thinking about that too, the `keep_bindings` seemed out of place with these changes but I wasn't sure of a approach, I agree with this.

File: packages/wrangler/src/index.tsx (hunk 1 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/index.tsx:532]: It seems rather clear to me, what could be expanded upon? I could just have context and am making assumption of knowledge here.

File: packages/wrangler/src/config/config.ts (hunk 2 above)
Comment (rozenmd) [reply packages/wrangler/src/config/config.ts:182]: That's what I intended, yeah. I assume we should aim for args/config parity?

File: packages/wrangler/src/publish.ts (hunk 3 above)
Comment (rozenmd) [reply packages/wrangler/src/publish.ts:507]: Yeah same lol

File: packages/wrangler/src/config/config.ts (hunk 2 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/config/config.ts:182]: We should

File: packages/wrangler/src/index.tsx (hunk 1 above)
Comment (petebacondarwin) [reply packages/wrangler/src/index.tsx:532]: Something like:

> Stop Wrangler from deleting vars that are not present in the wrangler.toml.
> By default Wrangler will remove all vars and replace them with those found in the `wrangler.toml` configuration.
> If your development approach is to modify vars after deployment via the dashboard you may wish to set this flag.

File: .changeset/grumpy-hornets-sit.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+"wrangler": patch
+---
+
+fix: rename keep_bindings to keep_vars, and make it opt-in, to keep wrangler.toml compatible with being used for Infrastructure as Code
+
+By default, wrangler.toml is the source of truth for your environment configuration, like a terraform file.
+
+If you change your settings (particularly your vars) in the dashboard, wrangler _will_ override them. If you want to disable this behavior, set this field to true.
```
Comment (petebacondarwin) [inline .changeset/grumpy-hornets-sit.md:9]: Is it worth mentioning that between 2.1.x(?) and this release the default was incorrectly set to the opposite?

Comment (rozenmd) [reply .changeset/grumpy-hornets-sit.md:9]: Good call

File: packages/wrangler/src/__tests__/publish.test.ts [hunk 5]
```diff
@@ -6540,6 +6540,54 @@ addEventListener('fetch', event => {});`
 			`A request to the Cloudflare API (/accounts/some-account-id/workers/services/test-name) failed`
 		);
 	});
+	describe("--keep-vars", () => {
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/publish.test.ts:6543]: Add a blank line here?

File: packages/wrangler/src/__tests__/publish.test.ts [hunk 6]
```diff
@@ -6540,6 +6540,54 @@ addEventListener('fetch', event => {});`
...
+
+			expect(std.out).toMatchInlineSnapshot(`
+			"Total Upload: xx KiB / gzip: xx KiB
+			Uploaded test-name (TIMINGS)
+			Published test-name (TIMINGS)
+			  https://test-name.test-sub-domain.workers.dev"
+		`);
+			expect(std.err).toMatchInlineSnapshot(`""`);
+		});
+		it("should not send keepVars by default", async () => {
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/publish.test.ts:6567]: And a blank line here?

File: packages/wrangler/src/publish.ts [hunk 7]
```diff
@@ -288,6 +289,7 @@ See https://developers.cloudflare.com/workers/platform/compatibility-dates for m
 
 	const jsxFactory = props.jsxFactory || config.jsx_factory;
 	const jsxFragment = props.jsxFragment || config.jsx_fragment;
+	const keepVars = props.keepVars || config.keep_vars;
```
Comment (petebacondarwin) [inline packages/wrangler/src/publish.ts:292]:
```suggestion
	const keepVars = props.keepVars ?? config.keep_vars;
```

That way `--keep-vars=false` would override `keep_vars = true` in wrangler.toml.

File: packages/wrangler/src/create-worker-upload-form.ts [hunk 8]
```diff
@@ -257,7 +258,7 @@ export function createWorkerUploadForm(worker: CfWorkerInit): FormData {
 		...(usage_model && { usage_model }),
 		...(migrations && { migrations }),
 		capnp_schema: bindings.logfwdr?.schema,
-		keep_bindings: ["plain_text", "json"],
+		...(keepVars && { keep_bindings: ["plain_text", "json"] }),
```
Comment (petebacondarwin) [inline packages/wrangler/src/create-worker-upload-form.ts:261]: I assume that not sending `keep_bindings` at all is valid? Have you tested this?
Is it safer to default to an empty array?

Comment (rozenmd) [reply packages/wrangler/src/create-worker-upload-form.ts:261]: Not sending `keep_bindings` is the old behaviour

**Pull request merged**